src/Microsoft.IdentityModel.Tokens/InMemoryCryptoProviderCache.cs

```diff
@@ -28,39 +26,56 @@ public InMemoryCryptoProviderCache() : this(new CryptoProviderCacheOptions())
-        /// <param name="options">Options used to create the event queue thread.</param>
-        /// <param name="tryTakeTimeout">The time used in ms for the timeout interval of the event queue. Defaults to 500 ms.</param>
         internal InMemoryCryptoProviderCache(CryptoProviderCacheOptions cryptoProviderCacheOptions, TaskCreationOptions options, int tryTakeTimeout = 500)
         {
-            if (cryptoProviderCacheOptions == null)
-                throw LogHelper.LogArgumentNullException(nameof(cryptoProviderCacheOptions));
-
+            _cryptoProviderCacheOptions = cryptoProviderCacheOptions ?? throw LogHelper.LogArgumentNullException(nameof(cryptoProviderCacheOptions));
             if (tryTakeTimeout <= 0)
                 throw LogHelper.LogArgumentException<ArgumentException>(nameof(tryTakeTimeout), $"{nameof(tryTakeTimeout)} must be greater than zero");
 
-            _cryptoProviderCacheOptions = cryptoProviderCacheOptions;
-            _signingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(cryptoProviderCacheOptions.SizeLimit, options, StringComparer.Ordinal, false) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
-            _verifyingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(cryptoProviderCacheOptions.SizeLimit, options, StringComparer.Ordinal, false) { OnItemRemoved = (SignatureProvider signatureProvider) => signatureProvider.CryptoProviderCache = null };
+            _signingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(
+                cryptoProviderCacheOptions.SizeLimit,
+                options,
+                comparer: StringComparer.Ordinal)
+            {
+                OnItemMovedToCompactedList = SetCryptoProviderCacheToNull,
+                OnItemRemovedFromCompactedList = DisposeSignatureProvider,
+                OnShouldRemoveFromCompactedList = IsCacheNullAndRefCountZero
+            };
+
+            _verifyingSignatureProviders = new EventBasedLRUCache<string, SignatureProvider>(
+                cryptoProviderCacheOptions.SizeLimit,
+                options,
+                comparer: StringComparer.Ordinal)
+            {
+                OnItemMovedToCompactedList = SetCryptoProviderCacheToNull,
+                OnItemRemovedFromCompactedList = DisposeSignatureProvider,
+                OnShouldRemoveFromCompactedList = IsCacheNullAndRefCountZero
+            };
+        }
+
+        internal CryptoProviderFactory CryptoProviderFactory { get; set; }
+
+        private static void DisposeSignatureProvider(SignatureProvider signatureProvider)
+        {
+            signatureProvider.Dispose();
+        }
+
+        private void SetCryptoProviderCacheToNull(SignatureProvider signatureProvider)
+        {
+            signatureProvider.CryptoProviderCache = null;
```

Comment:
Shouldn't we also set `signatureProvider.IsCached = false` here? Otherwise, we set it to true when provider is cached and never set it to false. So this `if` will never execute:
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/07a4b923e2448cdfb6c01841c87ede91cf8b6a98/src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs#L756

Reply:
@pmaytak SetCryptoProviderCacheToNull will only involve SignatureProviders that have been cached, they get disposed when the cache is compacted.

The code in CryptoProviderFactory will only be executed when a SignatureProvider is released AND it was never cached.

Because of the api on InMemoryCryptoProviderCache.TryRemove that will trigger SetCrytpriverCacheToNull when removed, i wanted to remember that the SignatureProvider was cached.

Now i remember why i had used the variable name 'wasChached' as i wanted to distinguish that the SignatureProvider was cached originally.